[styles] Generate global class names (#15140)
mui-org/material-ui: docs/src/pages/css-in-js/advanced/advanced.md

The class names are provided to the component with the `classes` variable.
The is particularly useful when styling nested elements in a component.

```jsx
// A style sheet
const useStyles = makeStyles({
  root: {}, // a style rule
  label: {}, // a nested style rule
});

export default function Nested(props) {
  const classes = useStyles(props); // { root: 'jss1', label: 'jss2' }
  return (
    <button className={classes.root}>
      <span className={classes.label}>
        nested
      </span>
    </button>
  );
}

export default function Parent() {
  return <Nested classes={{ label: 'my-label' }} />
}
```

However, the class names are often non deterministic, how can a parent component override the style of a nested element?

### withStyles

This is the simplest case, the wrapped component accepts a `classes` prop.
It simply merge the class names provided.

```jsx
const useStyles = withStyles({
  root: {},
  label: {},
})(props => (
  <button className={props.classes.root}>
    <span className={props.classes.label}> // 'jss2 my-label'
      nested
    </span>
  </button>

> ));

Author: It's closer to how people use styled components.

export default function Parent() {
  return <Nested classes={{ label: 'my-label' }} />
}
```

### makeStyles

The hook API requires a bit more work. You have to forward the parent props to the hook as a first argument.

```jsx
const useStyles = makeStyles({
  root: {},
  label: {},
});

export default function Nested(props) {

>   const classes = useStyles(props);

Reviewer: Having it take props allows for dynamic styling.

>   const classes = useStyles(props);

Author: @joshwooding Is correct. The first argument can be accessed in the style rules.

});
```

> 

Author: My hope is that people interested in how it works open the example in codesandbox.

> 

Reviewer 2: That's not how documentation works.

We have [an official plugin](https://github.com/hupe1980/gatsby-plugin-material-ui) that enables server-side rendering for `@material-ui/styles`.
Refer to the plugin's page for setup and usage instructions.

Refer to [this example](https://github.com/mui-org/material-ui/blob/next/examples/gatsby-next) for an up-to-date usage example.

### Next.js

You need to have a custom `pages/_document.js`, then copy [this logic](https://github.com/mui-org/material-ui/blob/next/examples/nextjs-next/pages/_document.js) to inject the server-side rendered styles into the `<head>` element.

Refer to [this example](https://github.com/mui-org/material-ui/blob/next/examples/nextjs-next) for an up-to-date usage example.

## Class names

The class names are generated by [the class name generator](/css-in-js/api/#creategenerateclassname-options-class-name-generator).

> ### Without a name

Author: I'm proposing `### Default`.

const className = `${productionPrefix}-${identifier}`;
```

### With `@material-ui/core`

The generated class names of the `@material-ui/core` components behave differently.
As soon as these conditions are met, the class names are **deterministic**.

- Only one theme provider is used: **no theme nesting**.
- The style sheet has a name and starts with `Mui`: all the Material-UI components.
- The `disableGlobal` option of the [class name generator](/css-in-js/api/#creategenerateclassname-options-class-name-generator) is `false`: the default.

These conditions are met with the most common use cases of `@material-ui/core`.

> It's almost like we were providing default values for the `classes` API.

Author: I have removed it.

For instance, this style sheet generates the following class names you can override:

```jsx
const useStyles = makeStyles({
  root: { /* … */ },
  label: { /* … */ },
  outlined: {
    /* … */
    '&$disabled': { /* … */ },
  },
  outlinedPrimary: {
    /* … */
    '&:hover': { /* … */ },
  },
  disabled: {},
}, { name: 'MuiButton' });
```

The CSS output:

```css
.MuiButton { /* … */ }
.MuiButton-label { /* … */ }
.MuiButton-outlined { /* … */ }
.MuiButton-outlined.disabled { /* … */ }
.MuiButton-outlinedPrimary: { /* … */ }
.MuiButton-outlinedPrimary:hover { /* … */ }

> ```

Author: I wanted to use snake-case as it's the convention. However, I think that it's simpler to use the component name and the style rule name without any transformation. People can move between the CSS and JavaScript domain more easily this way.

*This is a simplification of the `@material-ui/core/Button` component's style sheet.*

The customization the TextField can be cumbersome with the [`classes` API](#classes-prop)
when you have to provide each class name value manually.

> It's easier to use the default values we have described above:

Reviewer 2: ```suggestion
It's easier to use the default values, as described above. For example:
```

Repeat after me – there is no "we".

Author: there is no "we"

```jsx
import styled from 'styled-components';
import { TextField } from '@material-ui/core';

const StyledTextField = styled(TextField)`
  label.focused {
    color: green; 💚
  }
  .MuiOutlinedInput {
    fieldset {
      border-color: red; ❤️
    }
    &:hover fieldset {
      border-color: yellow; 💛
    }
    &.focused fieldset {
      border-color: green; 💚
    }
  }
`;
```

{{"demo": "pages/css-in-js/advanced/GlobalClassName.js"}}

## Global CSS

### `jss-plugin-global`

The [`jss-plugin-global`](#jss-plugins) plugin is installed in the default preset. You can use it to define global class names.

> {{"demo": "pages/css-in-js/advanced/GlobalCss.js"}}

Author: I can't think of anything else that won't be stitching.

### Hybrid

You can also combine JSS generated class names with global ones.

> {{"demo": "pages/css-in-js/advanced/HybridGlobalCss.js"}}

Author: I can't think of anything else that won't be stitching.
